### PR TITLE
Fixed #9107 -- Allowed URL parameters to set values of fields in admin inlines.

### PR DESCRIPTION
**Pull request by TapanGujjar**
ticket-9107

This is the most basic solution that I can think of for this feature. Please have a look at it @carltongibson

### CODE REVIEW

File: django/contrib/admin/options.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import json
 import operator
 import re
+from collections import defaultdict
```
Comment (carltongibson) [inline django/contrib/admin/options.py:5]: Run isort on this file.

File: django/contrib/admin/options.py [hunk 2]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
         Get the initial form data from the request's GET params.
         """
         initial = dict(request.GET.items())
+        formset_data = defaultdict(lambda: defaultdict(defaultdict))
+        formset_query_regex = r'inlinemode_set-(\d+)-(\d+)-(.*)'
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1510]: Compile the regular expression here as it'll be used multiple times below. Prefer `[0-9]` to `\d`. Expression should be anchored to avoid risk of partial matches. (Granted, you are using `re.match()` below, but it is always safer to be explicit.)
```python
formset_query_regex = re.compile(r'^inlinemode_set-([0-9]+)-([0-9]+)-(.*)$')
```

File: django/contrib/admin/options.py [hunk 3]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
         for k in initial:
             try:
+                match_obj = re.match(formset_query_regex, k)
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1513]: With the change above this becomes:
```python
match_obj = formset_query_regex.match(k)
```

File: django/contrib/admin/options.py [hunk 4]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
+                    inline_model_form_index = int(match_obj.group(2))
+                    inline_model_form_fieldname = match_obj.group(3)
+                    model_field = self.inlines[inline_model_index].model._meta.get_field(inline_model_form_fieldname)
+                    if isinstance(model_field, models.ManyToManyField):
+                        formset_data[inline_model_index][inline_model_form_index][inline_model_form_fieldname] \
+                            = initial.get(k).split(",")
+                    else:
+                        formset_data[inline_model_index][inline_model_form_index][inline_model_form_fieldname] \
+                            = initial.get(k)
+
```
Comment (carltongibson) [inline django/contrib/admin/options.py:1525]: Extract this logic to a helper function that can be tested independently.

File: django/contrib/admin/options.py [hunk 5]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
         initial = dict(request.GET.items())
+        formset_data = defaultdict(lambda: defaultdict(defaultdict))
+        formset_query_regex = r'inlinemode_set-(\d+)-(\d+)-(.*)'
         for k in initial:
             try:
+                match_obj = re.match(formset_query_regex, k)
+                if match_obj:
+                    inline_model_index = int(match_obj.group(1))
+                    inline_model_form_index = int(match_obj.group(2))
+                    inline_model_form_fieldname = match_obj.group(3)
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1517]: Note that you can use `match_obj[i]` instead of `match_obj.group(i)`.
See ticket-30116, #12892.

Comment (pope1ni) [reply django/contrib/admin/options.py:1517]: I also think that you could shorten these variable names. That would avoid the wrapping below and the backslash line continuation.

File: django/contrib/admin/options.py [hunk 6]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
+                    model_field = self.inlines[inline_model_index].model._meta.get_field(inline_model_form_fieldname)
+                    if isinstance(model_field, models.ManyToManyField):
+                        formset_data[inline_model_index][inline_model_form_index][inline_model_form_fieldname] \
+                            = initial.get(k).split(",")
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1521]: Single quotes please.

File: django/contrib/admin/options.py [hunk 7]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
+                    else:
+                        formset_data[inline_model_index][inline_model_form_index][inline_model_form_fieldname] \
+                            = initial.get(k)
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1524]: This is the same as above except for the `.split()`. Can you get the value, split if required and then assign?

File: django/contrib/admin/options.py [hunk 8]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
+
                 f = self.model._meta.get_field(k)
             except FieldDoesNotExist:
                 continue
+            except IndexError:
+                continue
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1530]: Is this tested? What if `fieldname` is invalid? That would raise a `KeyError` so that should be handled too?

File: django/contrib/admin/options.py [hunk 9]
```diff
@@ -1505,15 +1506,32 @@ def get_changeform_initial_data(self, request):
...
             # We have to special-case M2Ms as a list of comma-separated PKs.
             if isinstance(f, models.ManyToManyField):
                 initial[k] = initial[k].split(",")
-        return initial
+        return initial, formset_data
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1534]: This is changing the return type of `get_changeform_initial_data()` which is a documented function.

Documentation changes are missing. This may well also need a deprecation period unless there is an alternative way of solving this.

File: django/contrib/admin/options.py [hunk 10]
```diff
@@ -1590,9 +1608,12 @@ def _changeform_view(self, request, object_id, form_url, extra_context):
                 form_validated = False
         else:
             if add:
-                initial = self.get_changeform_initial_data(request)
+                initial, formset_initial_data = self.get_changeform_initial_data(request)
                 form = ModelForm(initial=initial)
                 formsets, inline_instances = self._create_formsets(request, form.instance, change=False)
+                # setting the initial value of the formset
+                for formset_index in range(len(formsets)):
+                    formsets[formset_index].initial = formset_initial_data[formset_index]
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1616]: I don't think the comment is necessary - it seems obvious what is happening.

You can also dispense with the index-based looping:
```python
for formset, data in zip(formsets, formset_initial_data):
    formset.initial = data
```

There is also an assumption that there is an entry in `formset_initial_data` for every formset. I guess that works because of the use of `defaultdict`.

File: django/contrib/admin/options.py [hunk 11]
```diff
@@ -1656,7 +1677,7 @@ def change_view(self, request, object_id, form_url='', extra_context=None):
         return self.changeform_view(request, object_id, form_url, extra_context)
 
     def _get_edited_object_pks(self, request, prefix):
-        """Return POST data values of list_editable primary keys."""
+        """Return POST data values of list_editable cmary keys."""
```
Comment (pope1ni) [inline django/contrib/admin/options.py:1680]: Please revert this unrelated change.

File: tests/admin_inlines/admin.py [hunk 12]
```diff
@@ -292,6 +293,21 @@ class TeacherAdmin(admin.ModelAdmin):
...
+class InlineTwoForHolder6Admin(admin.StackedInline):
+    model = InlineTwoForHolder6
+
+
+class Holder6Admin(admin.ModelAdmin):
+    inlines = [InlineOneForHolder6Admin, InlineTwoForHolder6Admin]
+
+
+# register for #9107
+site.register(Holder6, Holder6Admin)
```
Comment (pope1ni) [inline tests/admin_inlines/admin.py:310]: Are you able to reuse existing models? The names also don't seem very descriptive. What is "Holder6"?

File: tests/admin_inlines/tests.py [hunk 13]
```diff
@@ -1306,6 +1307,7 @@ def test_inline_formset_error_input_border(self):
                 '1px solid #ba2121',
             )
 
+    # test will fail
```
Comment (pope1ni) [inline tests/admin_inlines/tests.py:1310]: Does the test need fixing to work with the new behavior? Or is this change breaking expected behavior?

Comment (TapanGujjar) [reply tests/admin_inlines/tests.py:1310]: @pope1ni  @carltongibson  The reason is that as I mentioned in  [ticket-9107](https://code.djangoproject.com/ticket/9107), by default it will show 3 forms in the form set or the number of forms that are mentioned in the extra property in the inline model. 

In the test, the query parameter is trying to reference the 10th form in the inline formset which doesn't exist as only 3 forms will be shown, so here I have a dilemma, should I override the extra parameter of the inline model based on the query parameter or keep everything as it is?

One suggestion could be to have a parameter in the query to override how many forms for the inline formset to show, but that could make a simple feature more complex. 

@pope1ni  @carltongibson  any ideas on this?

File: tests/admin_inlines/admin.py (hunk 12 above)
Comment (TapanGujjar) [reply tests/admin_inlines/admin.py:310]: @pope1ni  I tried but I wanted to have a many to many field in the inline formset to test this feature. That is why I have created this model.

File: django/contrib/admin/options.py (hunk 9 above)
Comment (TapanGujjar) [reply django/contrib/admin/options.py:1534]: @carltongibson  @pope1ni  What I can do is to create a separate helper function as mentioned by @carltongibson  in the previous review and then call the function directly for the _changeform_view function. This will prevent the need to deprecate. What do you think ?

**Comment by felixxm** [comment]
Closing due to inactivity.
